packages/framework-core/src/services/event-store.ts

```diff
@@ -49,39 +50,58 @@ export class EventStore {
     entityName: string,
     entityID: UUID,
     pendingEnvelopes: Array<EventEnvelope>
-  ): Promise<EventEnvelope | null> {
+  ): Promise<Array<EventEnvelope>> {
     const logger = getLogger(this.config, 'EventStore#calculateAndStoreEntitySnapshot')
     logger.debug('Processing events: ', pendingEnvelopes)
     logger.debug(`Fetching snapshot for entity ${entityName} with ID ${entityID}`)
     const latestSnapshotEnvelope = await this.loadLatestSnapshot(entityName, entityID)
+    const snapshotsPerNewEntityId: Record<string, EventEnvelope> = {}
+    if (latestSnapshotEnvelope) {
+      snapshotsPerNewEntityId[latestSnapshotEnvelope.entityID.toString()] = latestSnapshotEnvelope
+    }
 
     logger.debug(
       `[EventStore#calculateAndStoreEntitySnapshot] Looking for the reducer for entity ${entityName} with ID ${entityID}`
     )
-    let newEntitySnapshot = latestSnapshotEnvelope
     for (const pendingEvent of pendingEnvelopes) {
-      newEntitySnapshot = await this.entityReducer(newEntitySnapshot, pendingEvent)
+      logger.debug(`Reducing snapshot for entity ${entityName} with ID ${pendingEvent.entityID.toString()}`)
+      const newEntitySnapshot = await this.entityReducer(
+        snapshotsPerNewEntityId[pendingEvent.entityID.toString()],
+        pendingEvent
+      )
+      snapshotsPerNewEntityId[newEntitySnapshot.entityID.toString()] = newEntitySnapshot
```

Comment:
@gonzalojaubert Wouldn't make sense to clarify this a bit, it's a bit confusing to calculate new snapshots for different entities than the `entityID` passed by param to this method.